### ISSUE TITLE
Process operations are abstracted on a per-platform basis

### ISSUE THREAD

**Issue by dguido** [facebook/osquery#1991]

**Comment by yying** [comment]
Currently, POSIX process related APIs are `ifdef`'d out of the code in order to focus on unit test executable generation.

**Comment by yying** [comment]
![screen shot 2016-04-19 at 2 54 37 pm](https://cloud.githubusercontent.com/assets/4278185/14657002/e23ee40e-0658-11e6-9151-842a02b51271.png)

We can now generate a `osquery_tests.exe` executable on Win64.

**Comment by yying** [comment]
Below is a list of process operations that will need to be abstracted across POSIX and Windows:

* **Launching a child process** (will need a method to deal with creating worker and extension processes, probably an enumeration)
* **Waiting for a process to terminate**
* **Terminating processes**
* **Obtaining the current process' PID**
* **Obtaining the current process' parent process PID** (difficult to obtain a process' parent process on Windows, we can emulate this by passing the parent process' PID as an environment variable/command line argument to the child)
* **Sleeping**
* **Signals** (for `osqueryd`, we can drop this since as a Windows Service, we are alerted to certain events such as request to stop service and imminent system shutdown. For `osqueryi`, we will need to find an alternative such as `SetConsoleCtrlHandler` to detect such events.)
* **Obtaining and setting environment variables**
* **Setting process priority**

**Comment by marpaia** [comment]
This has definitely been brought up before, but for posterity, when we were first making osquery, we took the code for the SQLite shell and ripped out all of `ifdef windows` code to trim it down / make it easier to understand and modify. Thus, if there's anything in `osquery/devtools/shell.cpp` that needs to be ported, it might be worth looking through the original SQLite shell code to see if an equivalent section exists in a more x-platform way (it probably does).

**Comment by yying** [comment]
Proposed process abstraction class interface.

<pre>
enum LaunchType {
    NORMAL_PROCESS = 0,
    WORKER_PROCESS,
    EXTENSION_PROCESS
};

class ProcessOperations
{
    public:
        bool launchProcess(std::string& exec_path, LaunchType type);
        void waitForProcess(pid_t pid);
        pid_t getProcessId();
        pid_t getParentProcessId();
        void sleep(unsigned int msec);
        bool hasEnvVar(const char *name);
        bool setEnvVar(const char *name, const char *value);
        bool getEnvVar(const char *name, char *value, unsigned int len);
        bool setProcessPrio(int who, int prio);
};
</pre>

Any further comments/suggestions?

**Comment by yying** [comment]
Updated interface:

```namespace osquery
{

namespace ProcessOperations 
{

#ifdef WIN32
using PlatformPidType = HANDLE;
#else
using PlatformPidType = pid_t;
#endif

const PlatformPidType kInvalidPid = (PlatformPidType) -1;

class PlatformProcess
{
  public:
    PlatformProcess(PlatformPidType id);
    ~PlatformProcess();

    PlatformPidType nativeHandle() { return id_; }
    
    bool kill();                 // TODO: consider making kill() return an enumeration that
                                 //       describes the various states
    bool setPriority(int prio);  // TODO: consider making prio an enumeration?
    int waitUntilExit();         // TODO: cosnsider making waitUntilExit() return an enumeration
                                 //       that describes the various states
    
    bool isValid() { return (id_ != kInvalidPid); }

    PlatformProcess& operator=(PlatformProcess& process);
    bool operator==(const PlatformProcess& process);
    bool operator!=(const PlatformProcess& process);
    
    static PlatformProcess launchWorker(const std::string& exec_path, const std::string& name);
    static PlatformProcess launchExtension(const std::string& exec_path, const std::string& extension);
    static PlatformProcess fromPlatformPid(PlatformPidType id);

  private:
    PlatformPidType id_ = kInvalidPid;
};

PlatformProcess getCurrentProcess();
PlatformProcess getLauncherProcess();

void sleep(unsigned int msec);

bool isLauncherProcessDead(PlatformProcess& launcher);
bool setEnvVar(const std::string& name, const std::string& value);
boost::optional<std::string> getEnvVar(const std::string& name);

}

}
```

**Comment by marpaia** [comment]
nit: there's no need to wrap any of this in a namespace other than the `osquery` namespace. we used to have a ton of namespaces and explicitly got rid of them, since it was more trouble than it was worth.

**Comment by yying** [comment]
Closed with #2069
